### ISSUE TITLE
SetComponents is sometimes very slow

### ISSUE THREAD

**Issue by iggolob** [artf/grapesjs#1909]
Hi all
We're working on a newsletter app and are using grapesjs as the email editor with the newsletter preset plguin and some custom changes. We store the HTMLs in our database and I noticed that some templates take a very long time to load with the editor.setComponents method. 

I tried to load the templates in the demo page through the import and while it loads almost instantly, I did notice a peculiar issue on the newsletter demo page ( https://grapesjs.com/demo-newsletter-editor.html ). If a component is selected before importing a HTML template it takes a lot longer to load (from 0.5s up to 5s). 

With inserting some logs I've figured that the backbone `set` method get's called a lot more. Mostly 'add' models get duplicated, but there are also a lot of 'change:style' models. But I'm kind of at a loss here what is necessary and what not :S
Maybe figuring out why setComponents get's a lot slower if a component is selected, could help me figure out why it is also slow in our project.

I've also noticed that this starts happening after release 0.14.40.
Any help is greatly appreciated.

**Comment by artf** [comment]
Hi @iggolob can you provide an example of a template and exact steps to reproduce the issue (talking about the fact is slower when you select a component)?

**Comment by iggolob** [comment]
Hi @artf 
The news letter demo page can be used to reproduce the issue ( https://grapesjs.com/demo-newsletter-editor.html ).
As a template you can use the preset HTML set for the demo page. Just copy the HTML from Export template option.
Now if you import that template it should be done almost immediately. But if you select any component, doesn't matter if image, text, cell or table and try to import, then it takes a noticeable long time, like a few seconds.
Hopefully this information is sufficient, I can also try to make a video, showing the steps.

**Comment by artf** [comment]
Yes I'm able to reproduce it, I've tried to give a quick look at the snapshot from the devtools but it's quite weird, definitely, we have to debug this issue better. At the first look the reason might be the Style Manager, which triggers an update on all properties, of the selected component, on any new component add.
A quick workaround would be to deselect a component on import
`editor.on('run:gjs-open-import-template', () => editor.select())`

**Comment by simplecommerce** [comment]
Hi, I am not sure if the issue is related or not, but I noticed this too.
In my case, I am using setComponents to reload pages I've edited.
Small pages are fine, but an example page we are building has over 3000 lines of code, takes almost 6 seconds to render.

The issue, is that it blocks all javascript on the page, so I can't even display a progress bar because it will be frozen.

If you have any ideas on how I could help debug this, I'd appreciate it, thanks!

**Comment by artf** [comment]
> Small pages are fine, but an example page we are building has over 3000 lines of code, takes almost 6 seconds to render.

@simplecommerce this is without a selected component?
Would be nice to have some data to test

**Comment by simplecommerce** [comment]
@artf I did a quick test with an empty page on my sandbox and it loads fine, I don't get it, most likely it has to be because of React and my other components on the page, but the funny thing is that, it's fine when there aren't too many lines of code to parse. I will try debugging some more to see if its really React or not.

If you have any ideas about that, I appreciate the feedback. 

Thanks for the reply!

**Comment by simplecommerce** [comment]
@artf Ok so it definitely seems to be something when integrating with React.

Here are two examples, one just normal JS, and the other in React.
You will be able to see both instances, one loads quickly and the other has a few seconds wait before it actually loads.

[HTML5](https://codesandbox.io/s/mj36lvr948)

[REACT](https://codesandbox.io/s/1rnq478nxl)

If you have any ideas, please let me know. I tried a couple of things with no luck but just minor improvements.

**Comment by artf** [comment]
@simplecommerce IMHO, based on your examples, it's more an issue with codesandbox transpiling process which takes that long, not the React itself. By the way, the best way to compare performance issues is to explore snapshots made via devtools

**Comment by simplecommerce** [comment]
@artf Hi, thanks for the reply, I understand what you mean about codesandbox, but what I meant was the loading when the actual editor is rendered, if you do the refresh and check, the editor itself once rendered, will parse the content, but you can actually see it freezes all white screen before it loads, which is the same result I had on my React instance, at first I thought it was in my code somewhere, but I did some tests and noticed that it is as soon as I call setComponents in my page with the 3k + lines of code, that it hangs (freezes) while executing the javascript code, and then resumes 5-6 seconds later. 

Can you tell me or show me what I would need to look at in the snapshots and print so you could see maybe what is bottling (if you can help at all).

Thanks for your response, I appreciate it!

**Comment by artf** [comment]
This shows you how to find issues with performances https://developers.google.com/web/tools/chrome-devtools/evaluate-performance/

**Comment by abozhinov** [comment]
HI @artf,
I have big problem with setComponents. I know the html is not so good :) but I want to know why in this case works so slow (around 1 min) 
https://www.dropbox.com/s/yk83qm312no2phq/test.html?dl=0

I load the whole website and set EDITABLE zone. The idea is user to see how the page will looks like.

**Comment by artf** [comment]
@abozhinov By using the webpage demo it took me a few seconds to import the template (for such a big template I think few seconds is a good deal) and after reload the template loads quite fast

**Comment by abozhinov** [comment]
@artf I've made a mistake send you wrong content. Check this html. I try to IMPORT it in https://grapesjs.com/demo.html but also freeze the CHROME :) 

https://www.dropbox.com/s/njebgew8qxae24u/test2?dl=0

**Comment by artf** [comment]
@abozhinov it took me less than a second to import... so I don't see any issue with your template.
Are you SELECTING SOMETHING before importing the template??

**Comment by abozhinov** [comment]
I use autoload with remote storage. The problem is that on some pages works fast but on others is super slow.
I will try to record a video.

**Comment by artf** [comment]
> I use autoload with remote storage. The problem is that on some pages works fast but on others is super slow.

again @abozhinov are you selecting something before importing the template? Because I've explained the reason here

**Comment by abozhinov** [comment]
No i don't select components.

**Comment by abozhinov** [comment]
Here i catch the HTML and add it.
  ``![image text](https://www.dropbox.com/s/i6vha7td9jjep1f/Screenshot%202019-06-15%20at%2015.29.25.png?dl=0)``

**Comment by abozhinov** [comment]
This is my structure:
[image text](https://www.dropbox.com/s/9dr9w4rfvs1pji3/Screenshot%202019-06-15%20at%2015.28.57.png?raw=1)

**Comment by artf** [comment]
mmm at first, remove `this.PageBuilder.render()` you don't need it, you just make it render twice. Then, before `this.PageBuilder.setComponents('...')` ensure no component is selected (`this.PageBuilder.select(null)`)

**Comment by abozhinov** [comment]
I add this but nothing change. I recorded video to see what's happen.
https://www.dropbox.com/s/yqn3nqmgwx1gwdm/test.mov?dl=0

**Comment by abozhinov** [comment]
This is the HTML
https://www.dropbox.com/s/s9dcwrbc465es8h/test?dl=0

**Comment by artf** [comment]
The last one takes me almost the same amount of time... and seems like those `<br>`s in the template increase the time. 
Anyway, for sure we could review and optimize the DOM traverser from some point but we are still talking about importing a big HTML template, so IMHO that time is not even that absurd. 
BTW you should import the HTML only once, after the edit you should only rely on the JSON which should be loaded more faster (in that case you skip the parser and DOM traverser)

**Comment by abozhinov** [comment]
The time now is OK. My problem was when parsing JSON 2 HTML. I forgot that have self closing tags like BR and my script generate <br></br>.
Thanks for your support.
